File: src/main/java/com/camackenzie/exvi/core/APIRequest.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStream;
import java.io.Reader;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublisher;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpRequest.Builder;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandler;
import java.net.http.HttpResponse.BodyHandlers;
import java.net.http.HttpResponse.BodySubscriber;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Flow;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 *
 * @author callum
 */
public class APIRequest<T> {

    private static final Gson gson = new Gson();

    private T body;
    private HashMap<String, String> headers;
    private String endpoint;

    public APIRequest(String endpoint, T body, HashMap<String, String> headers) {
        this.endpoint = endpoint;
        this.body = body;
        this.headers = headers;
    }

    public APIRequest(String endpoint, T body) {
        this(endpoint, body, new HashMap<>());
    }

    public APIRequest<T> withJsonHeader() {
        this.headers.put("content-type", "application/json");
        return this;
    }

    public T getBody() {
        return this.body;
    }

    public void setBody(T body) {
        this.body = body;
    }

    public HashMap<String, String> getHeaders() {
        return this.headers;
    }

    public void setHeaders(HashMap<String, String> headers) {
        this.headers = headers;
    }

    public String getEndpoint() {
        return this.endpoint;
    }

    public void setEndpoint(String e) {
        this.endpoint = e;
    }

    public Future<APIResult<T>> send() {
        HttpClient client = HttpClient.newHttpClient();
        Builder requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(this.endpoint));
        for (var header : this.headers.entrySet()) {
            requestBuilder.setHeader(header.getKey(),
                    header.getValue());
        }
        requestBuilder.POST(BodyPublishers.ofString(gson.toJson(this.body)));
        HttpRequest request = requestBuilder.build();

        return new APIResultFutureWrapper<T>((Class<T>) this.body.getClass(),
                client.sendAsync(request, BodyHandlers.ofString()));
    }

    public static <T> APIRequest<T> fromJson(String json) {
        return gson.fromJson(json, APIRequest.class);
    }

    public static <T> APIRequest<T> fromJson(Reader json) {
        return gson.fromJson(json, APIRequest.class);
    }

    public static <T> Future<APIResult<T>> send(String endpoint,
            T body,
            HashMap<String, String> headers) {
        return new APIRequest(endpoint, body, headers).send();
    }

    public static <T> Future<APIResult<T>> sendJson(String endpoint, T body) {
        return new APIRequest(endpoint, body).withJsonHeader().send();
    }

    private class APIResultFutureWrapper<T> implements Future<APIResult<T>> {

        private final transient Future<HttpResponse<String>> requestFuture;
        private final transient Class<T> tClass;
        private APIResult<T> result;

        public APIResultFutureWrapper(Class<T> tClass,
                CompletableFuture<HttpResponse<String>> sendAsync) {
            this.requestFuture = sendAsync;
            this.tClass = tClass;
        }

        @Override
        public boolean cancel(boolean mayInterruptIfRunning) {
            return this.requestFuture.cancel(mayInterruptIfRunning);
        }

        @Override
        public boolean isCancelled() {
            return this.requestFuture.isCancelled();
        }

        @Override
        public boolean isDone() {
            return this.requestFuture.isDone();
        }

        @Override
        public APIResult<T> get() throws InterruptedException, ExecutionException {
            return this.getFromInternalFutureGet(this.requestFuture.get());
        }

        @Override
        public APIResult<T> get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
            return this.getFromInternalFutureGet(this.requestFuture.get(timeout, unit));
        }

        private APIResult<T> getFromInternalFutureGet(HttpResponse<String> resp) {
            if (this.result == null) {
                String responseBody = resp.body();
                try {
                    // Format headers
                    HashMap<String, String> headers = new HashMap<>();
                    this.requestFuture.get()
                            .headers()
                            .map()
                            .entrySet()
                            .stream()
                            .forEach(en -> {
                                String headerVal = en.getValue().size() == 0 ? "" : en.getValue().get(0);
                                headers.put(en.getKey(), headerVal);
                            });
                    // Return proper api result
                    return new APIResult(this.requestFuture.get().statusCode(),
                            gson.fromJson(responseBody, this.tClass), headers);
                } catch (InterruptedException e) {
                    System.out.println(e);
                    Thread.currentThread().interrupt();
                } catch (ExecutionException e) {
                    System.err.println(e);
                }
                return null;
            } else {
                return this.result;
            }
        }

        public Class<T> getResultClass() {
            return this.tClass;
        }

    }

}

File: src/commonMain/kotlin/com/camackenzie/exvi/core/api/APIRequest.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.api

import com.camackenzie.exvi.core.util.SelfSerializable
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.statement.HttpResponse
import io.ktor.utils.io.core.*
import kotlinx.coroutines.*

/**
 *
 * @author callum
 */
@kotlinx.serialization.Serializable
class APIRequest<T : SelfSerializable> {
    val body: T

    @kotlinx.serialization.Transient
    var headers: HashMap<String, String> = HashMap()

    @kotlinx.serialization.Transient
    var endpoint: String = ""

    constructor(endpoint: String, body: T, headers: HashMap<String, String> = HashMap()) {
        this.endpoint = endpoint
        this.body = body
        this.headers = headers
    }

    constructor(other: APIRequest<*>, newBody: T) {
        endpoint = other.endpoint
        body = newBody
        headers = other.headers
    }

    fun withJsonHeader(): APIRequest<T> {
        headers["content-type"] = "application/json"
        return this
    }

    fun sendAsync(callback: (APIResult<String>) -> Unit) {
        CoroutineScope(Dispatchers.Default).launch {
            send(callback)
        }
    }

    fun sendAsync(callback: (HttpResponse, String) -> Unit) {
        CoroutineScope(Dispatchers.Default).launch {
            send(callback)
        }
    }

    suspend fun send(callback: (APIResult<String>) -> Unit) {
        send() { response, body ->
            val parsedResponse: APIResult<String> = APIResult(response.status.value, body, HashMap())
            callback(parsedResponse)
        }
    }

    suspend fun send(callback: (HttpResponse, String) -> Unit) = coroutineScope {
        launch {
            HttpClient {
                expectSuccess = false
            }.use { httpClient ->
                val reqHeaders = headers
                val reqBody = body.toJson()
                val response = httpClient.post<HttpResponse>(endpoint) {
                    headers {
                        for ((key, value) in reqHeaders) {
                            append(key, value)
                        }
                    }
                    body = reqBody
                }
                callback(response, response.receive())
            }
        }
    }

    companion object {
        suspend fun <T : SelfSerializable> request(
            endpoint: String,
            body: T,
            callback: (APIResult<String>) -> Unit
        ) {
            val req = APIRequest(endpoint, body)
            req.send() { result ->
                callback(result)
            }
        }

        fun <T : SelfSerializable> requestAsync(
            endpoint: String,
            body: T,
            callback: (APIResult<String>) -> Unit
        ) {
            val req = APIRequest(endpoint, body)
            req.sendAsync() { result ->
                callback(result)
            }
        }
    }

}
